src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java

```diff
@@ -0,0 +1,240 @@
+                    metric.PCT_T = (100.0 * firstReadTotalsByCycle[3][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_N = (100.0 * firstReadTotalsByCycle[4][i] / firstReadCountsByCycle[i]);
+                    metrics.addMetric(metric);
+                    firstReadLength = i;
+                }
+            }
+            if (seenSecondEnd) {
+                for (int i = 0; i < maxLengthSoFar; i++) {
+                    if (0 != secondReadCountsByCycle[i]) {
+                        final BaseDistributionByCycleMetrics metric = new BaseDistributionByCycleMetrics();
+                        metric.READ_END = 2;
+                        metric.CYCLE = (i + firstReadLength);
+                        metric.PCT_A = (100.0 * secondReadTotalsByCycle[0][i] / secondReadCountsByCycle[i]);
+                        metric.PCT_C = (100.0 * secondReadTotalsByCycle[1][i] / secondReadCountsByCycle[i]);
+                        metric.PCT_G = (100.0 * secondReadTotalsByCycle[2][i] / secondReadCountsByCycle[i]);
+                        metric.PCT_T = (100.0 * secondReadTotalsByCycle[3][i] / secondReadCountsByCycle[i]);
+                        metric.PCT_N = (100.0 * secondReadTotalsByCycle[4][i] / secondReadCountsByCycle[i]);
+                        metrics.addMetric(metric);
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    protected void runTool(final JavaSparkContext ctx) {
+        final JavaRDD<GATKRead> reads = getReads();
+        final MetricsFile<BaseDistributionByCycleMetrics, Integer> metricsFile = calculateBaseDistributionByCycle(reads);
+        saveResults(metricsFile, getHeaderForReads(), getReadSourceName());
+    }
+
+    /**
+     * Computes the MeanQualityByCycle. Creates a metrics file with relevant histograms.
+     */
+    public MetricsFile<BaseDistributionByCycleMetrics, Integer> calculateBaseDistributionByCycle(final JavaRDD<GATKRead> reads){
+        final JavaRDD<GATKRead> filteredReads = reads.filter(makeReadFilter(pfReadsOnly, alignedReadsOnly));
+        final HistogramGenerator hist = filteredReads.aggregate(new HistogramGenerator(),
+                (hgp, read) -> hgp.addRead(read),
+                (hgp1, hgp2) -> hgp1.merge(hgp2));
+
+        final MetricsFile<BaseDistributionByCycleMetrics, Integer> metricsFile = getMetricsFile();
+        hist.addToMetricsFile(metricsFile);
+        return metricsFile;
+    }
+
+    private static Function<GATKRead, Boolean> makeReadFilter(final boolean pf_read_only, final boolean aligned_reads_only) {
+        return read -> (!pf_read_only || !read.failsVendorQualityCheck()) &&
+                (!aligned_reads_only || !read.isUnmapped()) &&
+                !read.isSecondaryAlignment() &&
+                !read.isSupplementaryAlignment();
+    }
+
+    protected void saveResults(final MetricsFile<?, Integer> metrics, final SAMFileHeader readsHeader, final String inputFileName) {
```

Comment:
This also seems very similar to the last one I reviewed, maybe it could be moved to a library.